### PR TITLE
Add Ability to Write a BytesReference to Azure BlobContainer

### PR DESCRIPTION
**Pull request by fcofdez**
This pull request adds the ability to write a BytesReference to Azure and GCP BlobContainers.
This is a continuation of #66501

### CODE REVIEW

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java [hunk 1]
```diff
@@ -226,6 +227,11 @@ InputStream readBlob(String blobName, long position, long length) throws IOExcep
         }
     }
 
+    public void writeBlob(String blobName, BytesReference bytes, boolean failIfAlreadyExists) throws IOException {
+        final BlobInfo blobInfo = BlobInfo.newBuilder(bucketName, blobName).build();
+        executePutMultipartBlob(blobInfo, failIfAlreadyExists, BytesReference.toBytes(bytes));
```
Comment (original-brownbear) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:232]: `BytesReference.toBytes(bytes)` just copies all the bytes in most cases when working from anything that I'm afraid. Maybe we could use `com.google.cloud.storage.Storage#createFrom(com.google.cloud.storage.BlobInfo, java.io.InputStream, int, com.google.cloud.storage.Storage.BlobWriteOption...)` instead and use the stream we can get out of the `BytesReference` which won't do any copying?

**Review by original-brownbear — CHANGES_REQUESTED** [review]
Thanks for the follow-up Francisco + sorry for the delay in reviewing!

Azure looks just fine but I think GCS we should handle differently. The current approach doesn't really change anything relative to what we already do as far as I can tell, but there seem to be an easy fix as suggested inline.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java (hunk 1 above)
Comment (fcofdez) [reply plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:232]: Sadly `Storage#createFrom` perform multiples copies 🤦. The alternative, `Storage#writer(com.google.cloud.storage.BlobInfo, com.google.cloud.storage.Storage.BlobWriteOption...)` performs multiple copies too. I'm not sure if we should apply the optimization for GCP?

Comment (original-brownbear) [reply plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:232]: Urgh yea you're right GCP does crazy stuff there either way it seems. I'd say let's just fix Azure then and think about GCP at low priority in a follow up :)

Comment (fcofdez) [reply plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:232]: I'll revert the changes for GCP 👍

**Review by original-brownbear — APPROVED** [review]
LGTM, thanks Francisco!

**Comment by fcofdez** [comment]
Thanks for the review Armin!

**Pull request merged**